nav2_costmap_2d/test/integration/test_collision_checker.cpp

```diff
@@ -0,0 +1,328 @@
+#include "nav2_costmap_2d/testing_helper.hpp"
+#include "nav2_msgs/srv/get_robot_pose.hpp"
+#include "nav2_util/node_utils.hpp"
+#include "geometry_msgs/msg/pose_stamped.hpp"
+#include "tf2_ros/buffer.h"
+#include "tf2_ros/transform_listener.h"
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wpedantic"
+#include "tf2/utils.h"
+#pragma GCC diagnostic pop
+
+using namespace std::chrono_literals;
+using namespace std::placeholders;
+
+class RclCppFixture
+{
+public:
+  RclCppFixture() {rclcpp::init(0, nullptr);}
+  ~RclCppFixture() {rclcpp::shutdown();}
+};
+RclCppFixture g_rclcppfixture;
+
+class DummyCostmapSubscriber : public nav2_costmap_2d::CostmapSubscriber
+{
+public:
+  DummyCostmapSubscriber(
+    nav2_util::LifecycleNode::SharedPtr node,
+    std::string & topic_name)
+  : CostmapSubscriber(node, topic_name)
+  {}
+
+  void setCostmap(nav2_msgs::msg::Costmap::SharedPtr msg)
+  {
+    costmap_msg_ = msg;
+    costmap_received_ = true;
+  }
+};
+
+class DummyFootprintSubscriber : public nav2_costmap_2d::FootprintSubscriber
+{
+public:
+  DummyFootprintSubscriber(
+    nav2_util::LifecycleNode::SharedPtr node,
+    std::string & topic_name)
+  : FootprintSubscriber(node, topic_name)
+  {}
+
+  void setFootprint(geometry_msgs::msg::PolygonStamped::SharedPtr msg)
+  {
+    footprint_ = msg;
+    footprint_received_ = true;
+  }
+};
+
+class TestCollisionChecker : public nav2_util::LifecycleNode
+{
+public:
+  TestCollisionChecker(std::string name)
+  : LifecycleNode(name, "", true),
+    global_frame_("map")
+  {
+    // Declare non-plugin specific costmap parameters
+    declare_parameter("map_topic", rclcpp::ParameterValue(std::string("/map")));
+    declare_parameter("track_unknown_space", rclcpp::ParameterValue(true));
+    declare_parameter("use_maximum", rclcpp::ParameterValue(false));
+    declare_parameter("lethal_cost_threshold", rclcpp::ParameterValue(100));
+    declare_parameter("unknown_cost_value",
+      rclcpp::ParameterValue(static_cast<unsigned char>(0xff)));
+    declare_parameter("trinary_costmap", rclcpp::ParameterValue(true));
+  }
+
+  nav2_util::CallbackReturn
+  on_configure(const rclcpp_lifecycle::State & /*state*/)
+  {
+    RCLCPP_INFO(get_logger(), "Configuring");
+    tf_buffer_ = std::make_shared<tf2_ros::Buffer>(get_clock());
+    tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
+
+    std::string costmap_topic = "costmap_raw";
+    std::string footprint_topic = "published_footprint";
+
+    costmap_sub_ = std::make_shared<DummyCostmapSubscriber>(
+      shared_from_this(),
+      costmap_topic);
+
+    footprint_sub_ = std::make_shared<DummyFootprintSubscriber>(
+      shared_from_this(),
+      footprint_topic);
+
+    collision_checker_ = std::make_unique<nav2_costmap_2d::CollisionChecker>(
+      costmap_sub_, footprint_sub_, get_name());
+
+    get_robot_pose_service_ = rclcpp_node_->create_service<nav2_msgs::srv::GetRobotPose>(
```

Comment:
I think you should be able to create the service off of the main lifecycle node and therefore not require the rclcpp_node.

Reply:
I had tried to do that before, but it did not work since the service was hanging because I'm not actually spinning the lifecycle node here in this test. I used the `rclcpp_node` because it is spinning in a thread.

Reply:
Ok, np.